File: bp++/core/grammar.py
# $Id$
#
# Grammar productions.

import sys
import copy

from support import util

class Production(object):
    """Base class for all grammar productions. All productions are generally
    context-insensitive but can have semantic conditions attached to them.
    
    name: string - A name for this production; can be the None for anonymous
    productions.
    
    location: ~~Location - A location object decscribing the point of definition.
    """
    _counter = 0
    
    def __init__(self, name=None, location=None):
        self._name = name
        self._pred = None
        self._pre = []
        self._post = []
        self._location = location
        
        # Internal production name.
        Production._counter += 1
        self._tag = "T%d" % Production._counter
        
    def location(self):
        """Returns the location where the production was defined.
        
        Returns: ~~Location - The location.
        """
        return self._location
        
    def setPredicate(self, expr):
        """Adds a semantic Predicate to the production. The production will
        only be considered if the Predicate evaluates to True. 
        
        expr: ~~Expr - The expression to set a the predicate. It must evaluate
        to a boolen.
        """
        expr = expr.simplify()
        self._pred = expr
        
    def setName(self, name):
        """Sets the production's name.
        
        name: string - The name. Setting it to None turns the production into
        an anonymous one,
        """
        self._name = name
        
    def name(self):
        """Returns the production's name.
        
        Returns: string - The name, or None for anonymous productions. 
        """
        return self._name
    
    def predicate(self):
        """Returns the production's predicate.
        
        Returns: ~~Expr or None - The predicate, or None if none defined.
        """
        return self._pred

    def __str__(self):
        name = "($%s)" % self._name if self._name else ""
        pred = " if %s" % str(self._pred) if self._pred else ""
        return "%s%s%s" % (name, self._fmtDerived(), pred)
    
    def _fmtDerived(self):
        # To be overridden by derived classes.
        return "<general production>"

class Terminal(Production):
    """A terminal.
    
    type: ~~Type - The type of the literal.
    
    name: string - A name for this production; can be the None for anonymous
    productions.
    
    location: ~~Location - A location object decscribing the point of definition.
    """
    def __init__(self, type, name=None, location=None):
        super(Terminal, self).__init__(name, location=location)
        self._type = type

    def type(self):
        """Returns the type of the terminal.
        
        Returns: ~~Type - The type.
        """
        return self._type

    def _fmtDerived(self):
        return "<generic terminal>"
    
class Literal(Terminal):
    """A literal. A literal is anythig which can be directly scanned for as
    sequence of bytes. 
    
    literal: ~~Constant - The literal. 
    
    name: string - A name for this production; can be the None for anonymous
    productions.
    
    location: ~~Location - A location object decscribing the point of definition.
    """
    
    def __init__(self, literal, name=None, location=None):
        super(Literal, self).__init__(literal.type(), name)
        self._literal = literal

    def literal(self):
        """Returns the literal.
        
        Returns: ~~Constant - The literal.
        """
        return self._literal

    def _fmtDerived(self):
        return "literal('%s')" % self._literal

class Variable(Terminal):
    """A variable. A variable is terminal that will be parsed from the input
    stream according to its type, yet is not recognizable as such in advance
    by just looking at the available bytes. If we start parsing, we assume it
    will match (and if not, generate an error).
    
    type: ~~Type - The type of the variable.
    
    name: string - A name for this production; can be the None for anonymous
    productions.
    
    location: ~~Location - A location object decscribing the point of definition.
    """
    
    def __init__(self, type, name=None, location=None):
        super(Variable, self).__init__(type, name, location=location)

    def _fmtDerived(self):
        return "type(%s)" % self.type()

class EpsilonClass(Production):
    """The empty production.
    
    location: ~~Location - A location object decscribing the point of definition.
    """
    def __init__(self, location=None):
        super(EpsilonClass, self).__init__(None, location=location)
        self._tag = "epsilon"
        
    def _fmtDerived(self):
        return "<epsilon>"    

Epsilon = EpsilonClass()
    
class NonTerminal(Production):
    """A non-terminal.
    
    type: ~~Type - The type of the literal.
    
    name: string - A name for this production; can be the None for anonymous
    productions.
    
    location: ~~Location - A location object decscribing the point of definition.
    """
    def __init__(self, name=None, sym=None, location=None):
        super(NonTerminal, self).__init__(name, location=location)
        self._sym = sym
        if sym:
            self._tag = "%s_%s" % (sym, self._tag)

    def _fmtDerived(self):
        return "<generic non-terminal>"
    
class Sequence(NonTerminal):
    """A sequence of other productions. 
    
    seq: list of ~~Production - The sequence of productions. 
    
    name: string - A name for this production; can be the None 
    for anonymous productions.
    
    location: ~~Location - A location object decscribing the point of definition.
    """
    def __init__(self, seq, name=None, sym=None, location=None):
        super(Sequence, self).__init__(name, sym, location=location)
        self._seq = seq

    def sequence(self):
        """Returns the sequence of productions.
        
        Returns: list of ~~Production - The sequence.
        """
        return self._seq
    
    def addProduction(self, prod):
        """Appends one production to the sequence.
        
        prod: ~~Production - The production
        """
        self._seq += [prod]

    def _fmtDerived(self):
        return "( " + "; ".join([str(p) for p in self._seq]) + " )"
    
class Alternative(NonTerminal):
    """A alternative of other productions. 
    
    seq: list of ~~Production - The alternative of productions. 
    
    name: string - A name for this production; can be the None 
    for anonymous productions.
    
    location: ~~Location - A location object decscribing the point of definition.
    """
    def __init__(self, alt, name=None, sym=None, location=None):
        super(Alternative, self).__init__(name, sym, location=location)
        self._alt = alt

    def alternatives(self):
        """Returns the alternative of productions.
        
        Returns: list of ~~Production - The alternative.
        """
        return self._alt
    
    def addProduction(self, prod):
        """Appends one production to the alternative.
        
        prod: ~~Production - The production
        """
        self._alt += [prod]

    def _fmtDerived(self):
        return "( " + " || ".join([str(p) for p in self._alt]) + " )"

    
class Grammar:
    def __init__(self, name, root):
        """Instantiates a grammar given its root production.
        
        name: string - A name which uniquely identifies this grammar.
        root: Production - The grammar's root production.
        """
        self._name = name
        self._start = root._tag
        self._terms = {}
        self._prods = {}
        self._locations = {}
        self._nullable = {}
        self._first = {}
        self._follow = {}
        
        self._addProduction(root)
        #self._simplify()
        #self._simplify()
        self._computeTables()

    def name(self):
        """Returns the name of the grammar. The name uniquely identifies the
        grammar. 
        
        Returns: string - The name.
        """
        return self._name
        
    def check(self):
        """Checks the grammar for ambiguity. From an ambigious grammar, no
        parser can be generated.
        
        Returns: string or None - If the return valuue is None, the grammar is
        fine; otherwise the returned string contains a descriptions of the
        ambiguities encountered, suitable for inclusion in an error message.
        """
        
        def _loc(sym):
            try:
                return self._locations[sym]
            except KeyError:
                return str(sym)
        
        msg = ""
        
        for (sym, lookahead) in self._parser.items():
            for (term, rhss) in lookahead.values():
                if len(rhss) > 1:
                    msg += "%s: productions are ambigious for look-ahead symbol %s\n" % (_loc(sym), term)

            have_var = False
            total = 0
            for (term, rhss) in lookahead.values():
                if isinstance(term, Variable):
                    have_var = True
                total += 1
                
            if have_var and total > 1:
                msg += "%s: a variable must be the only option\n" % _loc(sym)
                
        return msg if msg != "" else None

    def parseTable(self):
        """Returns the grammar's parse table. The parse table has the form of
        a two-level dictionary. The first level is indexed by the LHS tag, and
        yield a second-level dictionary describing the alternatives we have
        when deriving a LHS. The alternative dictionary is indexed by the
        look-ahead symbol and yields the RHS as a list of (a) ~~Terminal
        instances for terminals, and (b) strings for non-terminals where the
        string is the LHS tag of the production to follow (the tag can be
        recursively used as the next index into the parse table). 
          
        The grammar must not be ambigious; if it is, this function will abort.
        Use ~~check to make sure the grammar is ok.
        
        Returns: dictionary - The dictionary as described above.
        
        Note: The dictionary returned by this function is different from the
        internal representation, which can map a look-ahead to multiple RHSs.
        In that case the grammar is ambigious though and will be flagged by
        ~~check.
        """
        assert not self.check()
        
        ptab = {}
        for (lhs, lookaheads) in self._parser.items():
            ptab[lhs] = dict([(sym, rhss[0]) for (sym, rhss) in lookaheads.values()])

        return ptab

    def startSymbol(self):
        """Returns the LHS symbol of the starting production. The tag can be
        used as index into the dictionary returned by ~~parseTable.
        
        Returns: string - The starting symbol.
        """
        return self._start
    
    @staticmethod
    def _isTerminal(p):
        return isinstance(p, Terminal)

    @staticmethod
    def _isEpsilon(p):
        return isinstance(p, EpsilonClass)
    
    @staticmethod
    def _isNonTerminal(p):
        return not (isinstance(p, Terminal) or isinstance(p, EpsilonClass))
        
    def _addProduction(self, p):
        if p._tag in self._prods:
            return

        self._locations[p._tag] = p.location() if p.location() else p._tag
        
        def _conv(p):
            return p._tag if Grammar._isNonTerminal(p) else p
        
        # We turn the productions into the classic lhs->rhs style here, with
        # alternatives being converted into multiple of them. We keep terminals
        # separately. For non-terminals, we only store their tag; for
        # Terminals the instance itself.
        if isinstance(p, Sequence):
            self._prods[p._tag] = [[_conv(q) for q in p.sequence()]]

            for q in p.sequence():
                self._addProduction(q)
            
        elif isinstance(p, Alternative):
            self._prods[p._tag] = [[_conv(q)] for q in p.alternatives()]

            for q in p.alternatives():
                self._addProduction(q)
                
        elif Grammar._isTerminal(p):
            pass

        elif Grammar._isEpsilon(p):
            pass
            
        else:
            util.internal_error("unknown case in Grammar.addProduction: %s" % str(p))
            
    def _simplify(self):
        # Remove some types of unecessary rules.
        for alts in self._prods.values():
            for i in range(len(alts)):
                p = alts[i]
            
                # For productions of the form "A->B; B->C;", let A points to C
                # directly.
                if len(p) == 1 and Grammar._isNonTerminal(p[0]):
                    succ = self._prods[p[0]]
                    if len(succ) == 1:
                        alts[i] = succ[0]
                        
                # For productions of the form "A-> xBy ; B -> terminal"',
                # replace B with the Terminal.
                for j in range(len(p)):
                    q = p[j]
                    if Grammar._isNonTerminal(q):
                        succ = self._prods[q]
                        if len(succ) == 1 and len(succ[0]) == 1 and not Grammar._isNonTerminal(succ[0][0]):
                            p[j] = succ[0][0]
            
        # Remove unused productions.
        def _closure(root, used):
            used.add(root)
            for a in self._prods[root]:
                for p in a:
                    if Grammar._isNonTerminal(p) and not p in used:
                        _closure(p, used)
                        
            return used
        
        used = _closure(self._start, set())
        
        for tag in set(self._prods) - used:
            del self._prods[tag]

        # Remove duplicates in the alternative list. We use the str()
        # representation to identify identical alternatives.
        for (tag, alts) in self._prods.items():
            news = []
            for a in alts:
                if str(a) not in [str(n) for n in news]:
                    news += [a]
                
            self._prods[tag] = news

    def _computeTables(self):
        # Computes FIRST, FOLLOW, & NULLABLE. This follows roughly the Algorithm
        # 3.13 from Modern Compiler Implementation in C by Appel/Ginsburg. See
        # http://books.google.com/books?id=A3yqQuLW5RsC&pg=PA49
        
        def _add(tbl, dst, src, changed):
            if tbl[dst] | set(src) == tbl[dst]:
                return changed
            
            tbl[dst] |= set(src)
            return True
            
        def _nullable(a, i, j):
            for l in range(i, j):
                if Grammar._isEpsilon(a[l]):
                    continue
                
                if Grammar._isTerminal(a[l]) or not self._nullable[a[l]]:
                    return False
            
            return True
        
        def _first(t):
            if Grammar._isEpsilon(t):
                return []
            
            if Grammar._isTerminal(t):
                return [t]
            
            return self._first[t]

        # Initialize sets.
        for t in self._prods:
            self._nullable[t] = False
            self._first[t] = set()
            self._follow[t] = set()

        # Iterate until no further change.
        while True:
            changed = False
            for (tag, alts) in self._prods.items():
                for a in alts:
                    k = len(a)

                    if _nullable(a, 0, k) and not self._nullable[tag]:
                        self._nullable[tag] = True
                        changed = True
                        
                    for i in range(k):
                        if _nullable(a, 0, i):
                            changed = _add(self._first, tag, _first(a[i]), changed)

                        if not Grammar._isNonTerminal(a[i]):
                            continue
                                
                        if _nullable(a, i+1, k):
                            changed = _add(self._follow, a[i], self._follow[tag], changed)

                        for j in range(i+1, k):
                            if _nullable(a, i+1, j):
                                changed = _add(self._follow, a[i], _first(a[j]), changed)
                                    
            if not changed:
                break
            
        # Build the parse table. 
        def _addla(la, term, rhs):
            assert Grammar._isTerminal(term)
            try:
                (term, cur) = la[str(term)]
                for c in cur:
                    if id(rhs) == id(c):
                        return

                cur += [rhs]
                    
                la[str(term)] = (term, cur)
                
            except KeyError:
                la[str(term)] = (term, [rhs])
 
        def _firstOfRhs(rhs):
            first = set()
            
            for sym in rhs:
                if Grammar._isEpsilon(sym):
                    continue
                
                if Grammar._isTerminal(sym):
                    return [sym]
                
                first |= self._first[sym]
                
                if not self._nullable[sym]:
                    break
            
            return first
                
        self._parser = {}
        
        for (tag, alts) in self._prods.items():
            lookahead = {}
            for rhs in alts:
                for term in _firstOfRhs(rhs):
                    _addla(lookahead, term, rhs)
                                
                if _nullable(rhs, 0, len(rhs)):
                    for term in self._follow[tag]:
                        _addla(lookahead, term, rhs)
                         
            self._parser[tag] = lookahead
        
    def __str__(self):
        s = "Grammar '%s'\n" % self._name
        for (tag, alts) in self._prods.items():
            for a in alts:
                start = "(*)" if tag == self._start else""
                prod = "; ".join([p if Grammar._isNonTerminal(p) else str(p) for p in a])
                s += "%3s %2s -> %s\n" % (start, tag, prod)
                
        s += "\n"
        s += "    Epsilon:\n"
        
        for (tag, eps) in self._nullable.items():
            s += "      %s = %s\n" % (tag, "true" if eps else "false" )
        
        s += "\n"
        s += "    First_1:\n"
        
        for (tag, first) in self._first.items():
            s += "      %s = %s\n" % (tag, [str(f) for f in first])
            
        s += "\n"
        s += "    Follow:\n"
        
        for (tag, follow) in self._follow.items():
            s += "      %s = %s\n" % (tag, [str(f) for f in follow])
            
        s += "\n"
        s += "    Parse table:\n"
        
        for (tag, lookahead) in self._parser.items():
            s += "      %s\n" % tag
            for (sym, rhss) in lookahead.values():
                s += "          %s -> #%d %s\n" % (sym, len(rhss), " || ".join([" ".join([str(r) for r in rhs]) for rhs in rhss]))
                
        return s

if __name__ == "__main__":

    from core import type
    from support import constant
    
    # Simple example grammar from
    #
    # http://www.cs.uky.edu/~lewis/essays/compilers/td-parse.html

    
    print "=== Example grammar 1"
    
    a = Literal(constant.Constant("a", type.Bytes()))
    b = Literal(constant.Constant("b", type.Bytes()))
    c = Literal(constant.Constant("c", type.Bytes()))
    d = Literal(constant.Constant("d", type.Bytes()))

#    A = Sequence([a])
#    B = Alternative([A, Sequence([A])])
#    print Grammar(B)
    
    
#    sys.exit(1)
    
    cC = Sequence([])
    bD = Sequence([])
    AD = Sequence([])
    aA = Sequence([])
    
    A = Alternative([Epsilon, a], sym="A")
    B = Alternative([Epsilon, bD], sym="B")
    C = Alternative([AD, b], sym="C")
    D = Alternative([aA, c], sym="D")

    ABA = Sequence([A,B,A])
    S = Alternative([ABA, cC], sym="S")

    cC.addProduction(c)
    cC.addProduction(C)
    
    bD.addProduction(b)
    bD.addProduction(D)

    AD.addProduction(A)
    AD.addProduction(D)
    
    aA.addProduction(a)
    aA.addProduction(A)

    g1 = Grammar("Grammar1", S)
    print str(g1)
    
    error = g1.check()
    if error:
        print error
    
    # Simple example grammar from "Parsing Techniques", Fig. 8.9
        
    print "=== Example grammar 2"

    hs = Literal(constant.Constant("#", type.Bytes()))
    pl = Literal(constant.Constant("(", type.Bytes()))
    pr = Literal(constant.Constant(")", type.Bytes()))
    no = Literal(constant.Constant("!", type.Bytes()))
    qu = Literal(constant.Constant("?", type.Bytes()))
    st = Variable(type.Bytes())

    F = Sequence([no, st], sym="Fact") 
    Q = Sequence([qu, st], sym="Question")
    
    S = Alternative([], sym="Session")
    SS = Sequence([pl, S, pr, S])
    Fs = Alternative([], sym="Facts")
    FsQ = Sequence([Fs, Q])
    FFs = Sequence([F,Fs])
    
    S.addProduction(FsQ)
    S.addProduction(SS)
    
    Fs.addProduction(FFs)
    Fs.addProduction(Epsilon)

    root = Sequence([S,hs], sym="Start")
    g2 = Grammar("Grammar2", root)
    
    print str(g2)
    
    error = g2.check()
    if error:
        print error
    
    # Simple example grammar from "Parsing Techniques", Fig. 8.9
        
    print "=== Example grammar 2"

    hdrkey = Variable(type.Bytes())
    colon = Literal(constant.Constant(":", type.Bytes()))
    ws = Literal(constant.Constant("[ \\t]*", type.Bytes()))
    hdrval = Variable(type.Bytes())
    nl = Literal(constant.Constant("[\\r\\n]", type.Bytes()))

    header = Sequence([hdrkey, ws, colon, ws, hdrval, nl], sym="Header")

    list1 = Sequence([header], sym="List1")
    list2 = Alternative([list1, Epsilon], sym="List2")
    list1.addProduction(list2)
    
    all = Alternative([list2, nl], sym="All")
    g3 = Grammar("Grammar3", all)
    
    print str(g3)
    
    error = g3.check()
    if error:
        print error

File: binpac/grammar.cc

#include <vector>

#include <util/util.h>

#include "grammar.h"
#include "expression.h"

using namespace binpac;

Grammar::Grammar(const string& name, shared_ptr<Production> root, const parameter_list& params, const parameter_list& attrs, const Location& l)
{
    _name = name;
    _root = root;
    _params = params;
    _attrs = attrs;
    _location = l;
    _epsilon = std::make_shared<production::Epsilon>();

    if ( root ) {
        _addProduction(root);
        _simplify();
        _simplify();
        _computeTables();
    }
}

const string& Grammar::name() const
{
    return _name;
}

const Location& Grammar::location() const
{
    return _location;
}

const Grammar::parameter_list& Grammar::parameters() const
{
    return _params;
}

string Grammar::_productionLocation(shared_ptr<Production> p) const
{
    string gloc, gname, ploc, pname;

    if ( _location != Location::None )
        gloc = util::fmt(" (%s)", string(_location).c_str());

    if ( p->location() != Location::None )
        ploc = util::fmt(" (%s)", string(p->location()).c_str());

    if ( _name.size() )
        gname = util::fmt("grammar %s%s, ", _name.c_str(), gloc.c_str());

    pname = util::fmt("production %s%s", p->symbol().c_str(), ploc.c_str());

    return util::fmt("%s%s", gname.c_str(), pname.c_str());
}

string Grammar::check() const
{
    string msg;

    for ( auto p : _lah_errors )
        msg += util::fmt("%s: look-ahead cannot depend on non-terminal\n", _productionLocation(p).c_str());

    for ( auto p : _nterms ) {
        auto lap = std::dynamic_pointer_cast<production::LookAhead>(p);

        if ( ! lap )
            continue;

        auto laheads = lap->lookAheads();

        std::set<string> syms1, syms2;

        for ( auto p : laheads.first )
            syms1.insert(util::fmt("%s (%s)", p->renderTerminal(), p->type()->render()));

        for ( auto p : laheads.second )
            syms2.insert(util::fmt("%s (%s)", p->renderTerminal(), p->type()->render()));

        auto defaults = lap->defaultAlternatives();

        if ( defaults.first && defaults.second ) {
            msg += util::fmt("%s: no look-ahead token for either alternative\n", _productionLocation(p).c_str());
            break;
        }

        if ( syms1.size() == 0 && syms2.size() == 0 ) {
            msg += util::fmt("no look-ahead symbol for either alternative in %s\n", _productionLocation(lap));
            break;
        }

        auto isect = util::set_intersection(syms1, syms2);

        if ( isect.size() )
            msg += util::fmt("%s is ambigious for look-ahead symbol(s) { %s }\n", _productionLocation(lap).c_str(), util::strjoin(isect, ", ").c_str());

        for ( auto q : util::set_union(laheads.first, laheads.second) ) {
            if ( ! std::dynamic_pointer_cast<production::Terminal>(q) )
                msg += util::fmt("%s: look-ahead cannot depend on non-terminal\n", _productionLocation(p).c_str());
                break;
        }
    }

    return msg;
}

const Grammar::production_map& Grammar::productions() const
{
    return _prods;
}

shared_ptr<Production> Grammar::root() const
{
    return _root;
}

string _fmtProds(std::set<string> prods)
{
    return util::strjoin(prods, ", ");
}

void Grammar::printTables(std::ostream& out, bool verbose)
{
    out << "=== Grammar " << _name << std::endl;

    for ( auto i : _prods ) {
        auto sym = i.first;
        auto prod = i.second;
        auto start = (prod == _root ? "(*)" : "");

        out << util::fmt(" %3s %s", start, prod->render().c_str()) << std::endl;
    }

    if ( ! verbose ) {
        out << std::endl;
        return;
    }

    out << std::endl;
    out << "  -- Epsilon:" << std::endl;

    for ( auto i : _nullable )
        out << "     " << i.first.c_str() << " = " << i.second << std::endl;

    out << std::endl;
    out << "  -- First_1:" << std::endl;

    for ( auto i : _first )
        out << "     " << i.first.c_str() << " = { " << _fmtProds(i.second).c_str() << " }" << std::endl;

    out << std::endl;
    out << "  -- Follow:" << std::endl;

    for ( auto i : _follow )
        out << "     " << i.first.c_str() << " = { " << _fmtProds(i.second).c_str() << " }" << std::endl;

    out << std::endl;
}

void Grammar::_addProduction(shared_ptr<Production> p)
{
    auto epsilon = std::dynamic_pointer_cast<production::Epsilon>(p);

    if ( epsilon ) {
        // We use our own internal instance for all epsilon productions.
        p = _epsilon;
    }

    if ( _prods_set.find(p) != _prods_set.end() )
        // Already added this.
        return;

    // Create a unique symbol.
    int cnt = 1;
    auto sym = p->symbol();

    while ( _prods.find(sym) != _prods.end() ) {
        sym = util::fmt("%s.%d", p->symbol(), ++cnt);
    }

    p->setSymbol(sym);
    _prods.insert(std::make_pair(sym, p));
    _prods_set.insert(p);

    auto nterm = std::dynamic_pointer_cast<production::NonTerminal>(p);

    if ( nterm ) {
        _nterms.push_back(nterm);

        for ( auto rhss : nterm->rhss() )
            for ( auto rhs : rhss )
                _addProduction(rhs);
    }
}

void Grammar::_computeClosure(shared_ptr<Production> root, std::set<string>* used)
{
    auto sym = root->symbol();

    if ( used->find(sym) != used->end() )
        return;

    used->insert(sym);

    auto nterm = std::dynamic_pointer_cast<production::NonTerminal>(root);

    if ( ! nterm )
        return;

    for ( auto rhss : nterm->rhss() )
        for ( auto rhs : rhss )
            _computeClosure(rhs, used);
}

void Grammar::_simplify()
{
    // Remove unused productions.

    std::set<string> used;
    _computeClosure(_root, &used);

    for ( auto sym : util::set_difference(util::map_keys(_prods), used) ) {
        _prods.erase(sym);

        for ( auto i = _nterms.begin(); i != _nterms.end(); i++ ) {
            if ( (*i)->symbol() == sym ) {
                _nterms.erase(i);
                break;
            }
        }
    }

    // Do production-specific simplyfing.
    //
    //
    for ( auto p : _prods )
        p.second->simplify();
}

bool Grammar::_add(std::map<string, symbol_set>* tbl, shared_ptr<Production> dst, const symbol_set& src, bool changed)
{
    auto idx = dst->symbol();
    auto t = tbl->find(idx);
    assert(t != tbl->end());

    auto set = t->second;
    auto union_ = util::set_union(set, src);

    if ( union_.size() == set.size() )
        // All in there already.
        return changed;

    (*tbl)[idx] = union_;

    return true;
}

bool Grammar::_isNullable(production_list::iterator i, production_list::iterator j)
{
    while ( i != j ) {
        auto rhs = *i++;

        if ( std::dynamic_pointer_cast<production::Epsilon>(rhs) )
            continue;

        if ( std::dynamic_pointer_cast<production::Terminal>(rhs) )
            return false;

        if ( ! _nullable[rhs->symbol()] )
            return false;
    }

    return true;
}

Grammar::symbol_set Grammar::_getFirst(shared_ptr<Production> p)
{
    if ( std::dynamic_pointer_cast<production::Epsilon>(p) )
        return symbol_set();

    if ( std::dynamic_pointer_cast<production::Terminal>(p) )
        return { p->symbol() };

    return _first[p->symbol()];
}

Grammar::symbol_set Grammar::_getFirstOfRhs(production_list rhs)
{
    auto first = symbol_set();

    for ( auto p : rhs ) {
        if ( std::dynamic_pointer_cast<production::Epsilon>(p) )
            continue;

        if ( std::dynamic_pointer_cast<production::Terminal>(p) )
            return { p->symbol() };

        first = util::set_union(first, _first[p->symbol()]);

        if ( ! _nullable[p->symbol()] )
            break;
    }

    return first;
}

void Grammar::_computeTables()
{
    // Computes FIRST, FOLLOW, & NULLABLE. This follows roughly the Algorithm
    // 3.13 from Modern Compiler Implementation in C by Appel/Ginsburg. See
    // http://books.google.com/books?id=A3yqQuLW5RsC&pg=PA49.

    // Initializde sets.
    for ( auto p : _nterms ) {
        auto sym = p->symbol();
        _nullable[sym] = false;
        _first[sym] = symbol_set();
        _follow[sym] = symbol_set();
    }

    // Iterator until no further change.
    while ( true ) {
        bool changed = false;

        for ( auto p : _nterms ) {
            auto sym = p->symbol();

            for ( auto rhss : p->rhss() ) {
                auto first = rhss.begin();
                auto last = rhss.end();

                if ( _isNullable(first, last) && ! _nullable[sym] ) {
                    _nullable[sym] = true;
                    changed = true;
                }

                for ( auto i = first; i != last; i++ ) {
                    auto rhs = *i;

                    if ( _isNullable(first, i) )
                        changed = _add(&_first, p, _getFirst(rhs), changed);

                    if ( ! std::dynamic_pointer_cast<production::NonTerminal>(rhs) )
                        continue;

                    auto next = i;
                    ++next;

                    if ( _isNullable(next, last) )
                        changed = _add(&_follow, rhs, _follow[sym], changed);

                    for ( auto j = next; j != last; ++j ) {
                        if ( _isNullable(next, j) ) {
                            changed = _add(&_follow, rhs, _getFirst(*j), changed);
                        }
                    }
                }
            }
        }

        if ( ! changed )
            break;
    }

    // Build the look-ahead sets.

    for ( auto p : _nterms ) {
        auto sym = p->symbol();
        auto lap = std::dynamic_pointer_cast<production::LookAhead>(p);

        if ( ! lap )
            continue;

        auto rhss = p->rhss();

        assert(rhss.size() == 2);
        auto r = rhss.begin();

        std::vector<symbol_set> laheads = { symbol_set(), symbol_set() };

        for ( auto i = 0; i < 2; ++i ) {
            auto rhs = *r++;

            for ( auto term : _getFirstOfRhs(rhs) )
                laheads[i] = util::set_union(laheads[i], { term });

            if ( _isNullable(rhs.begin(), rhs.end()) ) {
                for ( auto term : _follow[sym] )
                    laheads[i] = util::set_union(laheads[i], { term });
            }

        }

        production::LookAhead::look_aheads v0;
        production::LookAhead::look_aheads v1;

        std::set<string> lahs;

        for ( auto i = 0; i < 2; ++i ) {
            for ( auto s : laheads[i] ) {
                auto p = _prods.find(s);
                assert(p != _prods.end());
                auto t = std::dynamic_pointer_cast<production::Terminal>(p->second);

                if ( ! t ) {
                    _lah_errors.push_back(p->second);
                    break;
                }

                if ( i == 0 )
                    v0.insert(t);
                else
                    v1.insert(t);
            }
        }

        lap->setLookAheads(v0, v1);
    }
}
